File: whatsapp-doc-bot/services/file_handler.py
import io
import logging
from typing import Optional
from services.openai_service import read_image_content

logger = logging.getLogger(__name__)

SUPPORTED_IMAGE_TYPES = {"image/jpeg", "image/png", "image/gif", "image/webp", "image/bmp"}
SUPPORTED_DOC_TYPES = {
    "application/pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/msword",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "application/vnd.ms-excel",
    "text/plain",
    "text/csv",
}


def process_incoming_file(
    file_bytes: bytes,
    mime_type: str,
    filename: str = "",
) -> str:
    """
    Extrae texto del archivo según su tipo MIME.
    Retorna el texto extraído o una descripción del contenido.
    """
    mime_type = mime_type.lower().split(";")[0].strip()

    if mime_type == "application/pdf":
        return _extract_pdf(file_bytes)

    if mime_type in SUPPORTED_IMAGE_TYPES:
        return read_image_content(file_bytes, mime_type)

    if mime_type in (
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/msword",
    ):
        return _extract_docx(file_bytes)

    if mime_type in (
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "application/vnd.ms-excel",
    ):
        return _extract_xlsx(file_bytes)

    if mime_type in ("text/plain", "text/csv"):
        return _extract_text(file_bytes)

    logger.warning("Tipo MIME no soportado directamente: %s (%s)", mime_type, filename)
    return f"Archivo: {filename}. Tipo: {mime_type}. Contenido no extraíble automáticamente."


def _extract_pdf(file_bytes: bytes) -> str:
    try:
        import PyPDF2

        reader = PyPDF2.PdfReader(io.BytesIO(file_bytes))
        pages_text = []
        for i, page in enumerate(reader.pages):
            if i >= 20:  # límite de páginas para no exceder tokens
                break
            text = page.extract_text()
            if text:
                pages_text.append(text.strip())

        content = "\n\n".join(pages_text)
        if not content.strip():
            return "PDF sin texto extraíble (posiblemente escaneado)"
        logger.info("PDF extraído: %d caracteres de %d páginas", len(content), len(reader.pages))
        return content[:10000]
    except Exception as e:
        logger.error("Error extrayendo PDF: %s", e)
        return "No se pudo extraer el contenido del PDF"


def _extract_docx(file_bytes: bytes) -> str:
    try:
        from docx import Document

        doc = Document(io.BytesIO(file_bytes))
        paragraphs = [p.text.strip() for p in doc.paragraphs if p.text.strip()]

        # También extraer tablas
        for table in doc.tables:
            for row in table.rows:
                row_text = " | ".join(cell.text.strip() for cell in row.cells if cell.text.strip())
                if row_text:
                    paragraphs.append(row_text)

        content = "\n".join(paragraphs)
        logger.info("DOCX extraído: %d caracteres", len(content))
        return content[:10000] if content else "Documento Word vacío"
    except Exception as e:
        logger.error("Error extrayendo DOCX: %s", e)
        return "No se pudo extraer el contenido del documento Word"


def _extract_xlsx(file_bytes: bytes) -> str:
    try:
        import openpyxl

        wb = openpyxl.load_workbook(io.BytesIO(file_bytes), read_only=True, data_only=True)
        rows_text = []

        for sheet_name in wb.sheetnames:
            ws = wb[sheet_name]
            rows_text.append(f"[Hoja: {sheet_name}]")
            row_count = 0
            for row in ws.iter_rows(values_only=True):
                if row_count >= 100:  # límite de filas
                    break
                cells = [str(cell) for cell in row if cell is not None]
                if cells:
                    rows_text.append(" | ".join(cells))
                    row_count += 1

        content = "\n".join(rows_text)
        logger.info("XLSX extraído: %d caracteres", len(content))
        return content[:10000] if content else "Archivo Excel vacío"
    except Exception as e:
        logger.error("Error extrayendo XLSX: %s", e)
        return "No se pudo extraer el contenido del archivo Excel"


def _extract_text(file_bytes: bytes) -> str:
    try:
        for encoding in ("utf-8", "latin-1", "cp1252"):
            try:
                content = file_bytes.decode(encoding)
                logger.info("Texto extraído: %d caracteres (encoding: %s)", len(content), encoding)
                return content[:10000]
            except UnicodeDecodeError:
                continue
        return "Archivo de texto con encoding no compatible"
    except Exception as e:
        logger.error("Error extrayendo texto: %s", e)
        return "No se pudo leer el archivo de texto"


def get_file_extension(mime_type: str) -> str:
    extensions = {
        "application/pdf": ".pdf",
        "image/jpeg": ".jpg",
        "image/png": ".png",
        "image/gif": ".gif",
        "image/webp": ".webp",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document": ".docx",
        "application/msword": ".doc",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": ".xlsx",
        "application/vnd.ms-excel": ".xls",
        "text/plain": ".txt",
        "text/csv": ".csv",
    }
    mime_clean = mime_type.lower().split(";")[0].strip()
    return extensions.get(mime_clean, "")


def is_supported_file(mime_type: str) -> bool:
    mime_clean = mime_type.lower().split(";")[0].strip()
    return mime_clean in SUPPORTED_IMAGE_TYPES | SUPPORTED_DOC_TYPES

File: whatsapp-doc-bot/services/fileHandler.js
const { readImageContent } = require('./openai');

const SUPPORTED_IMAGE_TYPES = new Set([
  'image/jpeg',
  'image/png',
  'image/gif',
  'image/webp',
  'image/bmp',
]);

const SUPPORTED_DOC_TYPES = new Set([
  'application/pdf',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'application/msword',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  'application/vnd.ms-excel',
  'text/plain',
  'text/csv',
]);

/**
 * Extrae el texto/contenido de un archivo según su tipo MIME.
 * @param {Buffer} buffer  - Contenido del archivo
 * @param {string} mimetype
 * @param {string} filename
 * @returns {Promise<string>} Texto extraído
 */
async function processIncomingFile(buffer, mimetype, filename = '') {
  const mime = mimetype.toLowerCase().split(';')[0].trim();

  if (mime === 'application/pdf') {
    return extractPdf(buffer);
  }

  if (SUPPORTED_IMAGE_TYPES.has(mime)) {
    return readImageContent(buffer, mime);
  }

  if (
    mime === 'application/vnd.openxmlformats-officedocument.wordprocessingml.document' ||
    mime === 'application/msword'
  ) {
    return extractDocx(buffer);
  }

  if (
    mime === 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' ||
    mime === 'application/vnd.ms-excel'
  ) {
    return extractXlsx(buffer);
  }

  if (mime === 'text/plain' || mime === 'text/csv') {
    return extractText(buffer);
  }

  console.warn(`[fileHandler] Tipo MIME no soportado: ${mime} (${filename})`);
  return `Archivo: ${filename}. Tipo: ${mime}. Contenido no extraíble automáticamente.`;
}

async function extractPdf(buffer) {
  try {
    const pdfParse = require('pdf-parse');
    const data = await pdfParse(buffer);
    const text = data.text?.trim() || '';
    if (!text) return 'PDF sin texto extraíble (posiblemente escaneado)';
    console.log(`[fileHandler] PDF extraído: ${text.length} caracteres`);
    return text.slice(0, 10_000);
  } catch (err) {
    console.error('[fileHandler] Error extrayendo PDF:', err.message);
    return 'No se pudo extraer el contenido del PDF';
  }
}

async function extractDocx(buffer) {
  try {
    const mammoth = require('mammoth');
    const result = await mammoth.extractRawText({ buffer });
    const text = result.value?.trim() || '';
    console.log(`[fileHandler] DOCX extraído: ${text.length} caracteres`);
    return text.slice(0, 10_000) || 'Documento Word vacío';
  } catch (err) {
    console.error('[fileHandler] Error extrayendo DOCX:', err.message);
    return 'No se pudo extraer el contenido del documento Word';
  }
}

async function extractXlsx(buffer) {
  try {
    const XLSX = require('xlsx');
    const workbook = XLSX.read(buffer, { type: 'buffer' });
    const parts = [];

    for (const sheetName of workbook.SheetNames) {
      const sheet = workbook.Sheets[sheetName];
      parts.push(`[Hoja: ${sheetName}]`);

      // sheet_to_json con header:1 da arrays por fila
      const rows = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: '' });
      let rowCount = 0;
      for (const row of rows) {
        if (rowCount >= 100) break;
        const cells = row.filter((c) => c !== '').map(String);
        if (cells.length > 0) {
          parts.push(cells.join(' | '));
          rowCount++;
        }
      }
    }

    const text = parts.join('\n');
    console.log(`[fileHandler] XLSX extraído: ${text.length} caracteres`);
    return text.slice(0, 10_000) || 'Archivo Excel vacío';
  } catch (err) {
    console.error('[fileHandler] Error extrayendo XLSX:', err.message);
    return 'No se pudo extraer el contenido del archivo Excel';
  }
}

function extractText(buffer) {
  try {
    const text = buffer.toString('utf-8').trim();
    return text.slice(0, 10_000) || 'Archivo de texto vacío';
  } catch {
    try {
      return buffer.toString('latin1').trim().slice(0, 10_000);
    } catch (err) {
      console.error('[fileHandler] Error leyendo texto:', err.message);
      return 'No se pudo leer el archivo de texto';
    }
  }
}

function isSupportedFile(mimetype) {
  const mime = mimetype.toLowerCase().split(';')[0].trim();
  return SUPPORTED_IMAGE_TYPES.has(mime) || SUPPORTED_DOC_TYPES.has(mime);
}

function isImage(mimetype) {
  return SUPPORTED_IMAGE_TYPES.has(mimetype.toLowerCase().split(';')[0].trim());
}

module.exports = {
  processIncomingFile,
  isSupportedFile,
  isImage,
};
